src/main/java/com/vmware/vipclient/i18n/messages/service/LocaleService.java

```diff
@@ -4,28 +4,115 @@
  */
 package com.vmware.vipclient.i18n.messages.service;
 
-import java.util.HashMap;
-import java.util.List;
-import java.util.ListIterator;
-import java.util.Map;
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import java.util.*;
+import java.util.concurrent.Callable;
+import java.util.concurrent.FutureTask;
 
 import com.vmware.vipclient.i18n.VIPCfg;
 import com.vmware.vipclient.i18n.base.DataSourceEnum;
 import com.vmware.vipclient.i18n.base.cache.Cache;
 import com.vmware.vipclient.i18n.base.cache.FormatCacheItem;
+import com.vmware.vipclient.i18n.common.ConstantsMsg;
 import com.vmware.vipclient.i18n.messages.api.opt.server.RemoteLocaleOpt;
+import com.vmware.vipclient.i18n.messages.dto.LocaleDTO;
+import com.vmware.vipclient.i18n.util.FormatUtils;
 import com.vmware.vipclient.i18n.util.JSONUtils;
+import com.vmware.vipclient.i18n.util.LocaleUtility;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class LocaleService {
-
-    Logger                      logger        = LoggerFactory.getLogger(LocaleService.class.getName());
+    Logger logger = LoggerFactory.getLogger(LocaleService.class.getName());
     private static final String REGION_PREFIX = "region_";
     public static final String DISPN_PREFIX  = "dispn_";
+    private LocaleDTO dto;
 
     public LocaleService() {
+
+    }
+    public LocaleService(LocaleDTO dto) {
+        this.dto = dto;
+    }
+
+    public Map<String, String> getSupportedLanguages() {
+        return getSupportedLanguages(null, LocaleUtility.getFallbackLocales().iterator());
+    }
+
+    public Map<String, String> getSupportedLanguages(Iterator<Locale> fallbackLocalesIter) {
+        return getSupportedLanguages(null, fallbackLocalesIter);
```

Comment:
AS I said above,  I actually put the exact same code as suggestion in Huihui's PR, so the transfer to "ProductService" should be taken care of there.

Users won't directly use this method because this is Service layer so fallbackLocalesIter belong here.